oak_runtime/src/node/http/mod.rs

```diff
@@ -518,21 +494,122 @@ impl HttpResponseIterator {
         response_receiver.receive(&self.runtime)
     }
 
-    fn to_response(&self) -> Result<Response<Body>, http::Error> {
+    fn to_hyper_response(&self) -> Result<Response<Body>, OakStatus> {
         info!(
             "Generating response for runtime {} and reader {:?}.",
             self.runtime.node_id.0, self.response_reader
         );
         match self.read_response() {
-            Ok(http_response) => http::response::Builder::new()
-                .status(http_response.status as u16)
-                .body(Body::from(http_response.body)),
+            Ok(http_response) => to_hyper_http_response(http_response),
             Err(status) => {
                 error!("Could not read response: {}", status);
                 http::response::Builder::new()
                     .status(StatusCode::INTERNAL_SERVER_ERROR)
                     .body(Body::empty())
+                    .map_err(|err| {
+                        log::warn!("Could not build response: {}", err);
+                        OakStatus::ErrInternal
+                    })
             }
         }
     }
 }
+
+// Wrapper around a HashMap representing the headers in protobuf encapsulated HttpRequest and
+// HttpResponse. This struct is defined mainly to allow implementing `std::convert::From` for more
+// convenient conversion from hyper HeaderMaps.
+struct OakHeaderMap {
+    headers: HashMap<String, HeaderValue>,
+}
+
+impl OakHeaderMap {
+    // Flatten the headers into a vector of tuples (String, http::header::HeaderValue) that can
+    // be easily iterated over when populating the headers in the hyper Response builder.
+    fn to_vec(&self) -> Vec<(String, http::header::HeaderValue)> {
+        self.headers
+            .iter()
+            .flat_map(|item| {
+                item.1.values.iter().map(move |val| {
```

Comment:
I'm curious, why is the `move` necessary? Does it cause the converted value to reuse the underlying bytes buffer this way, instead of copying it?